middle_end/semantics_of_primitives.ml

```diff
@@ -141,6 +141,8 @@ let for_primitive (prim : Clambda_primitives.primitive) =
   | Pdls_get ->
       (* only read *)
       No_effects, No_coeffects
+  | Ppoll ->
+      Arbitrary_effects, Has_coeffects
```

Comment:
Other allocating primitives in Flambda use No_effects (or Only_generative_effects when they return a fresh value whose identity is distinct from previous values). This is justified in the following documentation comment in the .mli file:

```
Exceptions arising from allocation points, for example "out of memory" or
    exceptions propagated from finalizers or signal handlers, are treated as
    "effects out of the ether" and thus ignored for our determination here
    of effectfulness.
```

I would thus expect to see `No_effects, No_coeffects` here: if we have decided to ignore asynchronous behaviors, this function is just the identity at the unit type.

Reply:
I agree in principle, but a primitive that has no effects and does not return any useful value will just get removed.
Adding arbitrary effects is not ideal but it will prevent Flambda from removing it.
In a way, you can see `Ppoll` as including the semantics of `Popaque`, for similar reasons.

Reply:
(This suggests that maybe the effect analysis of Flambda could be extended with "asynchronous effects".)

Reply:
Yes, but I don't see the point. What would be the semantics ? If it implies arbitrary effects, then we can never remove such primitives, which means that the compiler is forbidden from removing any allocation.

Reply:
I originally put `No_effects, No_coeffects` here but then (a) I noticed that when this primitive was present previously (on the multicore branch), it was [changed](https://github.com/ctk21/ocaml-multicore/commit/3e009f1be1682c76e0bd755af183151a17bf66de) specifically to `Arbitrary_effects, Has_coeffects`, presumably for the reason that @lthls identifies.

Reply:
... and (b) I ran a very simple test and it seemed not to be doing a poll, so I went looking for culprits.

Reply:
It would be nice to have a comment that explains the reasoning here -- there is a contradiction between the documented semantics and the actual choices.

@lthls I'm not sure, but just idle thoughts.

1. There is a difference between allocations and explicit poll points. Typical allocations *may* run asynchronous effects but I don't necessarily *intend* them to. Explicit poll points, on the other hand, should act as a barrier for anything that can observe asynchronous effects.
2. It would be possible to decouple "allocations" and "implicit poll points" in the intermediate representation, where allocations are pure (in a real sense, not in the "we ignore asynchronous code" sense). The poll insertion pass also inserts implicit poll points (without allocations). Implicit poll points may be moved around, as long as we preserve the guarantee that we run one of them "often enough". On the other hand, in practice placing an implicit poll point on an allocation is free (they use the same young_limit check), while separating them means twice the number of checks.

Reply:
I've added a comment.

Reply:
The comment is `(* Effects to prevent poll from being optimised away *)` but I think the reason is stronger. I agree with @gasche's analysis, in particular one might want to reason that polling happens wherever the source code claims (not meaningful now, but it would be important in combination with masking where polling would be the only way to run actions in a mask).